<span style="color:red">WIP. Markdown tips:  https://www.markdownguide.org/basic-syntax/#</span>

<h1><center> 4AB00. Intermediate assignment</center></h1>
<h2><center> Modeling and simulation of an autonomous vehicle storage and retrieval system</center></h2>
    
# 1 Subject

Today's competitive environment, in which deliveries need to be faster and order sizes become smaller, forces material handling providers to progressively develop new and better solutions. A recent development in automated material-handling technology for unit load storage and retrieval is the autonomous vehicle storage and retrieval system (AVS/RS). Figure 1 shows a representation of an AVS/RS for the handling of totes, while Figure 2 illustrates a single tier (floor, level). The storage racks are single-deep and double-sided. Each storage position is of the same size and can hold one tote. Lifts are mounted at fixed locations at one end of each storage aisle. The input/output (I/O) point is located at the first tier beside each lift. Autonomous vehicles are dedicated to a storage aisle within a specific tier (so-called tier-captive configuration). The first position on either side of the storage aisle in all tiers serves as a buffer and is used to manage the transfer of totes between vehicles and lifts (see Figure 2). One buffer (out) handles totes which have been retrieved, the other one (in), located on the other side of the storage aisle, handles the totes to be stored. The presence of these buffers allows the lift and vehicle to work independently of each other. 

The throughput performance of AVS/RS systems can be affected by design decisions (such as number of tiers, number of aisles, and depth of aisles) as well as operational decisions (such as tote storage location and order assignment).

Figure 1: AVS/RS with tier-captive configuration [1] | Figure 2: Single tier in AVS/RS [1]
- | - 
<img src="./Course_4DC10/nickpy-exercises/figures/avs.png" alt="drawing" width="400"/> | <img src="./Course_4DC10/nickpy-exercises/figures/avrs2.png" alt="drawing" width="400"/>




# 2 Goal
The goal of this assignment is to study this new material handling solution by using computer simulation. This study can be restricted to the retrieval process only, i.e., the vehicles perform only single retrieval cycles. Clearly, the retrieval phase is the most critical activity from an organisational viewpoint, as it is directly related to customer service level and &mdash; in contrast to the storage phase &mdash; it cannot be postponed to a period of low workload. 

The objectives of this assignment can be summarized as follows:
- Develop a simulation model of one aisle that consists of Level tiers with retrieval process to assess the system performance in terms of throughput and flow time.
- Systematically develop the model process by process, and verify and validate each process using analytical results.
- Investigate design trade-offs, i.e., via different layouts (number of columns and tiers), for this material handling system.

# 3 Description of the AVS/RS
The AVS/RS is a fully automated system that can store and retrieve unit loads. The unit loads are placed in totes, a type of box, and these totes are placed in a designated aisle and column for storage. When retrieved, an autonomous vehicle places the totes in a buffer at the end of the aisle. There is one autonomous vehicle for every aisle within the tier. This is called the tier-captive configuration. From the buffer, the totes are picked by a lift that serves all tiers inone aisle, see Figure 2. In this assignment only one aisle that consists of `Level` tiers has to be modelled. Each aisle has depth number of columns of width `dv` metres and height `dl` metres.

Requests for a tote arrive for each tier, asking for a tote with a random (uniformly distributed) column assigned to it. The inter arrival time of requests is exponentially distributed with mean value `arrive`.

The vehicles transport totes along the aisle (one vehicle per tier). The time taken by the vehicle to process an order depends on the tote's location. The vehicle has to travel to this location, load the tote (it takes a fixed amount of time equal to `lv` seconds), then go back to the end of the aisle and, finally, unload the tote to the buffer (again this takes `lv` seconds). The speed profile of each vehicle is described as follows. First, the vehicle goes with a constant acceleration `av` *m/s$^2$* until it reaches its maximum velocity `vmaxv`, then it moves with this velocity and decelerates with a constant deceleration `av` to approach its destination point. If the tote is located too close to the end of the aisle, it is possible that the vehicle will not reach its maximum speed.

The vehicles unload the totes to a location where it is picked up by the lift. At any given time only one tote per tier is allowed to occupy such a position.

The lift transports the totes from every tier to the ground floor. The processing time depends on the tote's location and the fixed time to load and unload the tote (`ll` seconds for loading and `ll` seconds for unloading). The maximum speed of the lift is `vmaxl` *m/s* with a speed profile similar to that of the vehicles (constant acceleration/deceleration of `al` *m/s<sup>2</sup>*).

# 4 Modeling of the AVS/RS
The model consists of several processes communicating via channels, see Figure 3.

| Figure 3: The processes in the model | 
- 
<img src="./Course_4DC10/nickpy-exercises/figures/avrs.png" alt="drawing" width="400"/> 

A brief description of the processes is given below:
- **G:** The generator models the arrival of orders for totes. There is one generator per tier.
- **DB:** The demand buffer stores the orders to be picked up by the vehicles. The vehicle processes the orders under first-come-first-serve policy. It can always receive a tote, but it releases a tote only when the vehicle is ready to pick it up.
- **V:** The vehicle can be modeled as a server with a variable processing time.
- **GDV:** The process consisting of a Generator, Demand Buffer and Vehicle for one tier of the AVS/RS system.
- **B:** The buffer stores the totes from the vehicles awaiting for a pick-up by the lift. The buffer operates under first-come-first-serve policy and has a finite capacity of `bc` totes per tier. So, every tier has a buffer location of finite capacity of `bc` totes. All those locations together form the buffer B that accommodates all the totes which are processed by the same lift. When the buffer is not empty (there is at least one tote on one of the tiers), the buffer sends information about the tier of the longest waiting tote to the lift upon request. This information is used by the lift to determine the location of the tote (the number of the tier the tote is located in). Once the lift is ready to pick it up, the tote is released to the lift. There are two types of queues in the buffer: `xs` contains all the totes to be picked up by the lift and there are Level counters `n` counting the totes from the corresponding tier.
- **L:** The lift, similar to the process V, is modeled as a machine with a variable processing time. A lift first receives a destination tier. Then the process delays for time required for the lift to go to the tote location and pick up the tote. Then it receives the tote, and next it delays for the time required to go to the ground floor and to unload the tote. After that the lift process is ready to release the tote to the Exit process.
- **E:** In the exit process all the data related to each tote are collected. This process calculates the average throughput and the average flow-time. The number of totes to be processed by the AVS/RS during one simulation run is determined by the constant `Number_of_orders`.

# 5 Remark concerning report
The report has to contain concise answers to the questions posed. If code has to be presented, present the
specification only of the modified process. Avoid unnecessary long explanations, however answers to analytical
questions without an argument will result in no points.


<span style="color:red">(Dit stuk moet natuurlijk nog aangepast worden)</span>

# 6 Assignment
When making a SimPy specification, one does not build an entire model and then starts debugging. A model is build step by step. During each step proper functionality is checked. We therefore model one process at a time, and connect the process with a Generator and/or Exit process to test proper functionality. In order to test proper functionality we first use analytical methods to determine the expected outcome. Then we run simulations to verify if these outcomes are indeed obtained. If that is the case, we can be more convinced that the process has been modeled correctly.

First, we initialize our simulation


In [3]:
%reset -f
from numpy import random
import simpy
import math
import numpy
numpy.seterr(divide='ignore')
from Course_4DC10.simChi.core.channel_select import Channel, process, selected, Environment
from dataclasses import dataclass

## 6.1 Exercise 1: Generator (5 points)
So the first step is to write a proper generator for a tier. To that end, the first model we build consists of (only) a Generator process and an Exit process collecting the totes generated by the Generator. The inter arrival time of requests is exponentially distributed with mean `arrive`. The requested tote should be retrieved from a random (uniformly distributed) column, where there are `depth` columns in total.

a. Determine (analytically) the throughput and flow time that should result from your model with only the generator process and the exit process.

Answer: <input type="text" id="6.1.a" name="6.1.a"/> 

<span style="color:red">(werkt nog niet)</span>


b. Complete the simulation code from the following template:

In [4]:

# =================================
# # Global variables
# =================================
arrive = 70.0             # inter arrival time of requests
depth  = 55               # the number of columns
number_of_orders = 10000  # the number of orders to process

# =================================
# # Tote definition
# =================================
@dataclass
class Tote:
    entrytime: float = 0.0
    column: int = 0

# =================================
# # Generator definition
# =================================
def Generator(env, sending_channel):
    ### BEGIN SOLUTION
    ta_dist = lambda: random.exponential(arrive)
    column_dist = lambda:  random.randint(1, depth)
    ### END SOLUTION
    x = Tote()
    while True:
        x.entrytime = env.now
        # x.column = ..... 
        ### BEGIN SOLUTION
        x.column = column_dist()
        ### END SOLUTION
        send = sending_channel.send(x)
        yield env.select(send)
        # delay = ..... 
        ### BEGIN SOLUTION
        delay = env.timeout(ta_dist())
        ### END SOLUTION
        yield delay


# =================================
# # Exit definition
# =================================
def Exit(env, receiving_channel):
    x = Tote()
    mphi = 0
    for i in range(1, number_of_orders + 1):
        receive = receiving_channel.receive()
        x = yield env.select(receive)
        mphi = (i - 1) / i * mphi + (env.now - x.entrytime) / i
        print("tote = %d; Entrytime = %10.4f; Column = %2d; Mean throughput = %8.6f; Mean flowtime = %6.4f\n" 
            %(i , x.entrytime, x.column, numpy.float64(i) / env.now, mphi ) )


# =================================
# # Main
# =================================
env = Environment()
a = Channel(env)
G = env.process(Generator(env, a))
E = env.process(Exit(env, a))
env.run(until=E)
print ("simulation has ended")


tote = 1; Entrytime =     0.0000; Column = 33; Mean throughput =      inf; Mean flowtime = 0.0000

tote = 2; Entrytime =    45.0308; Column = 10; Mean throughput = 0.044414; Mean flowtime = 0.0000

tote = 3; Entrytime =   139.5073; Column = 49; Mean throughput = 0.021504; Mean flowtime = 0.0000

tote = 4; Entrytime =   173.7481; Column =  3; Mean throughput = 0.023022; Mean flowtime = 0.0000

tote = 5; Entrytime =   185.7169; Column =  8; Mean throughput = 0.026923; Mean flowtime = 0.0000

tote = 6; Entrytime =   188.1694; Column =  8; Mean throughput = 0.031886; Mean flowtime = 0.0000

tote = 7; Entrytime =   218.1403; Column =  4; Mean throughput = 0.032089; Mean flowtime = 0.0000

tote = 8; Entrytime =   254.4899; Column = 24; Mean throughput = 0.031435; Mean flowtime = 0.0000

tote = 9; Entrytime =   307.5849; Column =  9; Mean throughput = 0.029260; Mean flowtime = 0.0000

tote = 10; Entrytime =   371.4131; Column = 51; Mean throughput = 0.026924; Mean flowtime = 0.0000

tote = 11


tote = 1240; Entrytime = 85647.3555; Column = 35; Mean throughput = 0.014478; Mean flowtime = 0.0000

tote = 1241; Entrytime = 85895.2804; Column = 35; Mean throughput = 0.014448; Mean flowtime = 0.0000

tote = 1242; Entrytime = 85904.0549; Column = 35; Mean throughput = 0.014458; Mean flowtime = 0.0000

tote = 1243; Entrytime = 85905.4719; Column = 27; Mean throughput = 0.014469; Mean flowtime = 0.0000

tote = 1244; Entrytime = 85922.7370; Column =  8; Mean throughput = 0.014478; Mean flowtime = 0.0000

tote = 1245; Entrytime = 85962.1425; Column = 30; Mean throughput = 0.014483; Mean flowtime = 0.0000

tote = 1246; Entrytime = 86013.3617; Column =  6; Mean throughput = 0.014486; Mean flowtime = 0.0000

tote = 1247; Entrytime = 86036.5924; Column = 53; Mean throughput = 0.014494; Mean flowtime = 0.0000

tote = 1248; Entrytime = 86045.1651; Column = 11; Mean throughput = 0.014504; Mean flowtime = 0.0000

tote = 1249; Entrytime = 86117.1235; Column = 31; Mean throughput = 0.014504; Mea


tote = 2188; Entrytime = 153313.2273; Column = 48; Mean throughput = 0.014271; Mean flowtime = 0.0000

tote = 2189; Entrytime = 153336.4265; Column = 11; Mean throughput = 0.014276; Mean flowtime = 0.0000

tote = 2190; Entrytime = 153362.1867; Column = 33; Mean throughput = 0.014280; Mean flowtime = 0.0000

tote = 2191; Entrytime = 153388.3406; Column = 33; Mean throughput = 0.014284; Mean flowtime = 0.0000

tote = 2192; Entrytime = 153392.5267; Column = 30; Mean throughput = 0.014290; Mean flowtime = 0.0000

tote = 2193; Entrytime = 153508.8121; Column = 29; Mean throughput = 0.014286; Mean flowtime = 0.0000

tote = 2194; Entrytime = 153676.8775; Column = 15; Mean throughput = 0.014277; Mean flowtime = 0.0000

tote = 2195; Entrytime = 153759.8365; Column = 49; Mean throughput = 0.014276; Mean flowtime = 0.0000

tote = 2196; Entrytime = 153818.0617; Column = 33; Mean throughput = 0.014277; Mean flowtime = 0.0000

tote = 2197; Entrytime = 153848.8693; Column =  4; Mean throughput = 0.0

tote = 3459; Entrytime = 242539.3731; Column = 19; Mean throughput = 0.014262; Mean flowtime = 0.0000

tote = 3460; Entrytime = 242556.2614; Column = 26; Mean throughput = 0.014265; Mean flowtime = 0.0000

tote = 3461; Entrytime = 242610.1127; Column = 10; Mean throughput = 0.014266; Mean flowtime = 0.0000

tote = 3462; Entrytime = 242805.8405; Column =  9; Mean throughput = 0.014258; Mean flowtime = 0.0000

tote = 3463; Entrytime = 242873.5896; Column = 36; Mean throughput = 0.014258; Mean flowtime = 0.0000

tote = 3464; Entrytime = 242897.1373; Column = 21; Mean throughput = 0.014261; Mean flowtime = 0.0000

tote = 3465; Entrytime = 242950.0036; Column = 50; Mean throughput = 0.014262; Mean flowtime = 0.0000

tote = 3466; Entrytime = 242955.0430; Column = 14; Mean throughput = 0.014266; Mean flowtime = 0.0000

tote = 3467; Entrytime = 243081.7367; Column = 48; Mean throughput = 0.014263; Mean flowtime = 0.0000

tote = 3468; Entrytime = 243156.3225; Column = 45; Mean throughput = 0.01

tote = 4766; Entrytime = 338214.8050; Column =  4; Mean throughput = 0.014092; Mean flowtime = 0.0000

tote = 4767; Entrytime = 338283.9517; Column = 41; Mean throughput = 0.014092; Mean flowtime = 0.0000

tote = 4768; Entrytime = 338377.0223; Column = 35; Mean throughput = 0.014091; Mean flowtime = 0.0000

tote = 4769; Entrytime = 338455.8851; Column = 13; Mean throughput = 0.014090; Mean flowtime = 0.0000

tote = 4770; Entrytime = 338522.3791; Column =  3; Mean throughput = 0.014091; Mean flowtime = 0.0000

tote = 4771; Entrytime = 338540.8627; Column = 34; Mean throughput = 0.014093; Mean flowtime = 0.0000

tote = 4772; Entrytime = 338570.3373; Column =  4; Mean throughput = 0.014095; Mean flowtime = 0.0000

tote = 4773; Entrytime = 338655.3255; Column = 40; Mean throughput = 0.014094; Mean flowtime = 0.0000

tote = 4774; Entrytime = 338850.7595; Column = 26; Mean throughput = 0.014089; Mean flowtime = 0.0000

tote = 4775; Entrytime = 338873.8184; Column = 39; Mean throughput = 0.01


tote = 5848; Entrytime = 415650.4872; Column = 24; Mean throughput = 0.014070; Mean flowtime = 0.0000

tote = 5849; Entrytime = 415706.9373; Column = 43; Mean throughput = 0.014070; Mean flowtime = 0.0000

tote = 5850; Entrytime = 415720.9970; Column = 13; Mean throughput = 0.014072; Mean flowtime = 0.0000

tote = 5851; Entrytime = 415805.4965; Column =  3; Mean throughput = 0.014071; Mean flowtime = 0.0000

tote = 5852; Entrytime = 415818.4874; Column = 19; Mean throughput = 0.014073; Mean flowtime = 0.0000

tote = 5853; Entrytime = 415853.0875; Column = 48; Mean throughput = 0.014075; Mean flowtime = 0.0000

tote = 5854; Entrytime = 416059.8757; Column = 27; Mean throughput = 0.014070; Mean flowtime = 0.0000

tote = 5855; Entrytime = 416087.4595; Column = 47; Mean throughput = 0.014072; Mean flowtime = 0.0000

tote = 5856; Entrytime = 416115.1065; Column = 42; Mean throughput = 0.014073; Mean flowtime = 0.0000

tote = 5857; Entrytime = 416121.3194; Column = 15; Mean throughput = 0.0


tote = 6577; Entrytime = 465287.7538; Column = 49; Mean throughput = 0.014135; Mean flowtime = 0.0000

tote = 6578; Entrytime = 465392.4868; Column = 17; Mean throughput = 0.014134; Mean flowtime = 0.0000

tote = 6579; Entrytime = 465486.4874; Column = 23; Mean throughput = 0.014134; Mean flowtime = 0.0000

tote = 6580; Entrytime = 465593.9228; Column = 10; Mean throughput = 0.014132; Mean flowtime = 0.0000

tote = 6581; Entrytime = 465646.2061; Column = 12; Mean throughput = 0.014133; Mean flowtime = 0.0000

tote = 6582; Entrytime = 465843.5651; Column = 12; Mean throughput = 0.014129; Mean flowtime = 0.0000

tote = 6583; Entrytime = 465848.2614; Column = 10; Mean throughput = 0.014131; Mean flowtime = 0.0000

tote = 6584; Entrytime = 465922.3002; Column = 41; Mean throughput = 0.014131; Mean flowtime = 0.0000

tote = 6585; Entrytime = 465926.5930; Column = 28; Mean throughput = 0.014133; Mean flowtime = 0.0000

tote = 6586; Entrytime = 466045.0439; Column = 19; Mean throughput = 0.0


tote = 7521; Entrytime = 529967.5079; Column = 27; Mean throughput = 0.014191; Mean flowtime = 0.0000

tote = 7522; Entrytime = 530014.9453; Column = 12; Mean throughput = 0.014192; Mean flowtime = 0.0000

tote = 7523; Entrytime = 530116.6583; Column = 38; Mean throughput = 0.014191; Mean flowtime = 0.0000

tote = 7524; Entrytime = 530376.5162; Column = 28; Mean throughput = 0.014186; Mean flowtime = 0.0000

tote = 7525; Entrytime = 530378.3126; Column = 39; Mean throughput = 0.014188; Mean flowtime = 0.0000

tote = 7526; Entrytime = 530435.7099; Column = 28; Mean throughput = 0.014188; Mean flowtime = 0.0000

tote = 7527; Entrytime = 530676.1342; Column = 52; Mean throughput = 0.014184; Mean flowtime = 0.0000

tote = 7528; Entrytime = 530694.5342; Column = 13; Mean throughput = 0.014185; Mean flowtime = 0.0000

tote = 7529; Entrytime = 530705.5928; Column =  7; Mean throughput = 0.014187; Mean flowtime = 0.0000

tote = 7530; Entrytime = 530706.0494; Column = 51; Mean throughput = 0.0


tote = 8517; Entrytime = 598772.8716; Column = 40; Mean throughput = 0.014224; Mean flowtime = 0.0000

tote = 8518; Entrytime = 598859.7889; Column = 52; Mean throughput = 0.014224; Mean flowtime = 0.0000

tote = 8519; Entrytime = 598955.7252; Column = 34; Mean throughput = 0.014223; Mean flowtime = 0.0000

tote = 8520; Entrytime = 599103.8527; Column =  7; Mean throughput = 0.014221; Mean flowtime = 0.0000

tote = 8521; Entrytime = 599186.9955; Column = 14; Mean throughput = 0.014221; Mean flowtime = 0.0000

tote = 8522; Entrytime = 599276.5415; Column = 38; Mean throughput = 0.014220; Mean flowtime = 0.0000

tote = 8523; Entrytime = 599346.8733; Column = 20; Mean throughput = 0.014220; Mean flowtime = 0.0000

tote = 8524; Entrytime = 599448.6314; Column = 17; Mean throughput = 0.014220; Mean flowtime = 0.0000

tote = 8525; Entrytime = 599477.5477; Column = 32; Mean throughput = 0.014221; Mean flowtime = 0.0000

tote = 8526; Entrytime = 599502.7657; Column = 24; Mean throughput = 0.0


tote = 9516; Entrytime = 665194.3593; Column = 21; Mean throughput = 0.014306; Mean flowtime = 0.0000

tote = 9517; Entrytime = 665276.7502; Column = 32; Mean throughput = 0.014305; Mean flowtime = 0.0000

tote = 9518; Entrytime = 665282.8684; Column = 43; Mean throughput = 0.014307; Mean flowtime = 0.0000

tote = 9519; Entrytime = 665297.2329; Column =  1; Mean throughput = 0.014308; Mean flowtime = 0.0000

tote = 9520; Entrytime = 665420.2399; Column = 51; Mean throughput = 0.014307; Mean flowtime = 0.0000

tote = 9521; Entrytime = 665439.3469; Column =  6; Mean throughput = 0.014308; Mean flowtime = 0.0000

tote = 9522; Entrytime = 665470.5138; Column = 25; Mean throughput = 0.014309; Mean flowtime = 0.0000

tote = 9523; Entrytime = 665518.1165; Column = 51; Mean throughput = 0.014309; Mean flowtime = 0.0000

tote = 9524; Entrytime = 665660.8882; Column = 10; Mean throughput = 0.014308; Mean flowtime = 0.0000

tote = 9525; Entrytime = 665710.3826; Column = 25; Mean throughput = 0.0

c. Confirm that your simulations produce the correct throughput and flow time. Also determine the minimal and maximal column returned by your simulation.

Run your simulation at least 10 times. What is the range of outcomes for the mean throughput? Determine (roughly, by trial and error) the number of orders required to get good estimates for the throughput (first three non-zero digits are correct).

## 6.2 Exercise 2: Demand Buffer (5 points)
Next, we include the Demand buffer which is modelled as a standard FIFO buffer with infinite capacity.

a. Determine (analytically) the throughput and flow time that should result from your model with only the generator process, demand buffer, and the exit process.

b. Complete the simulation code in the template below.

Note: The previously defined variables and functions remain unchanged, only the Demand Buffer is added, and the main code is changed.

In [5]:
# =================================
# # Demand Buffer definition
# =================================
def Demand_Buffer(env, receiving_channel, sending_channel):
    xs = [] # list of totes
    while True:
            ### BEGIN SOLUTION
            receiving = receiving_channel.receive() 
            sending = sending_channel.send(xs[0]) if len(xs)>0 else None
            yield env.select(sending,receiving)
            if selected(sending):
                xs = xs[1:]
            if selected(receiving):
                x = receiving.received_entity
                xs = xs + [x]
            ### END SOLUTION

# =================================
# # Main
# =================================
env = Environment()
a = Channel(env)
b = Channel(env)
G = env.process(Generator(env, a))
D = env.process(Demand_Buffer(env, a,b))
E = env.process(Exit(env, b))
env.run(until=E)
print ("simulation has ended")

tote = 1; Entrytime =     0.0000; Column =  3; Mean throughput =      inf; Mean flowtime = 0.0000

tote = 2; Entrytime =    39.8242; Column = 53; Mean throughput = 0.050221; Mean flowtime = 0.0000

tote = 3; Entrytime =    54.5709; Column = 15; Mean throughput = 0.054974; Mean flowtime = 0.0000

tote = 4; Entrytime =   100.6343; Column = 26; Mean throughput = 0.039748; Mean flowtime = 0.0000

tote = 5; Entrytime =   116.1352; Column = 33; Mean throughput = 0.043053; Mean flowtime = 0.0000

tote = 6; Entrytime =   158.1627; Column = 46; Mean throughput = 0.037936; Mean flowtime = 0.0000

tote = 7; Entrytime =   159.1645; Column = 16; Mean throughput = 0.043980; Mean flowtime = 0.0000

tote = 8; Entrytime =   185.8103; Column = 30; Mean throughput = 0.043055; Mean flowtime = 0.0000

tote = 9; Entrytime =   212.1326; Column =  7; Mean throughput = 0.042426; Mean flowtime = 0.0000

tote = 10; Entrytime =   536.3377; Column = 34; Mean throughput = 0.018645; Mean flowtime = 0.0000

tote = 11


tote = 922; Entrytime = 62846.7735; Column = 38; Mean throughput = 0.014671; Mean flowtime = 0.0000

tote = 923; Entrytime = 62964.0254; Column =  5; Mean throughput = 0.014659; Mean flowtime = 0.0000

tote = 924; Entrytime = 63063.7888; Column = 54; Mean throughput = 0.014652; Mean flowtime = 0.0000

tote = 925; Entrytime = 63140.5786; Column =  3; Mean throughput = 0.014650; Mean flowtime = 0.0000

tote = 926; Entrytime = 63201.8777; Column = 16; Mean throughput = 0.014651; Mean flowtime = 0.0000

tote = 927; Entrytime = 63271.0345; Column = 16; Mean throughput = 0.014651; Mean flowtime = 0.0000

tote = 928; Entrytime = 63337.9173; Column = 32; Mean throughput = 0.014652; Mean flowtime = 0.0000

tote = 929; Entrytime = 63408.6039; Column = 28; Mean throughput = 0.014651; Mean flowtime = 0.0000

tote = 930; Entrytime = 63448.0476; Column =  3; Mean throughput = 0.014658; Mean flowtime = 0.0000

tote = 931; Entrytime = 63470.7813; Column = 51; Mean throughput = 0.014668; Mean flowtime


tote = 1353; Entrytime = 92966.7915; Column = 52; Mean throughput = 0.014554; Mean flowtime = 0.0000

tote = 1354; Entrytime = 92976.1166; Column = 27; Mean throughput = 0.014563; Mean flowtime = 0.0000

tote = 1355; Entrytime = 92984.0092; Column = 15; Mean throughput = 0.014572; Mean flowtime = 0.0000

tote = 1356; Entrytime = 93046.9305; Column = 49; Mean throughput = 0.014573; Mean flowtime = 0.0000

tote = 1357; Entrytime = 93186.3338; Column = 27; Mean throughput = 0.014562; Mean flowtime = 0.0000

tote = 1358; Entrytime = 93216.9264; Column = 19; Mean throughput = 0.014568; Mean flowtime = 0.0000

tote = 1359; Entrytime = 93273.3773; Column =  5; Mean throughput = 0.014570; Mean flowtime = 0.0000

tote = 1360; Entrytime = 93323.7429; Column =  8; Mean throughput = 0.014573; Mean flowtime = 0.0000

tote = 1361; Entrytime = 93348.2438; Column = 24; Mean throughput = 0.014580; Mean flowtime = 0.0000

tote = 1362; Entrytime = 93398.9767; Column =  2; Mean throughput = 0.014583; Mea


tote = 1948; Entrytime = 133544.5282; Column = 47; Mean throughput = 0.014587; Mean flowtime = 0.0000

tote = 1949; Entrytime = 133771.3196; Column = 51; Mean throughput = 0.014570; Mean flowtime = 0.0000

tote = 1950; Entrytime = 133913.4026; Column = 41; Mean throughput = 0.014562; Mean flowtime = 0.0000

tote = 1951; Entrytime = 133951.0574; Column = 34; Mean throughput = 0.014565; Mean flowtime = 0.0000

tote = 1952; Entrytime = 133991.7732; Column = 19; Mean throughput = 0.014568; Mean flowtime = 0.0000

tote = 1953; Entrytime = 134074.7283; Column = 36; Mean throughput = 0.014567; Mean flowtime = 0.0000

tote = 1954; Entrytime = 134194.2925; Column = 50; Mean throughput = 0.014561; Mean flowtime = 0.0000

tote = 1955; Entrytime = 134218.9447; Column = 34; Mean throughput = 0.014566; Mean flowtime = 0.0000

tote = 1956; Entrytime = 134298.1306; Column = 50; Mean throughput = 0.014565; Mean flowtime = 0.0000

tote = 1957; Entrytime = 134304.8604; Column = 31; Mean throughput = 0.0


tote = 2817; Entrytime = 195007.0950; Column = 24; Mean throughput = 0.014446; Mean flowtime = 0.0000

tote = 2818; Entrytime = 195095.5868; Column = 36; Mean throughput = 0.014444; Mean flowtime = 0.0000

tote = 2819; Entrytime = 195239.0933; Column = 48; Mean throughput = 0.014439; Mean flowtime = 0.0000

tote = 2820; Entrytime = 195260.7591; Column = 29; Mean throughput = 0.014442; Mean flowtime = 0.0000

tote = 2821; Entrytime = 195269.3160; Column =  9; Mean throughput = 0.014447; Mean flowtime = 0.0000

tote = 2822; Entrytime = 195292.6212; Column = 51; Mean throughput = 0.014450; Mean flowtime = 0.0000

tote = 2823; Entrytime = 195296.8088; Column = 33; Mean throughput = 0.014455; Mean flowtime = 0.0000

tote = 2824; Entrytime = 195402.8395; Column = 30; Mean throughput = 0.014452; Mean flowtime = 0.0000

tote = 2825; Entrytime = 195563.0391; Column = 30; Mean throughput = 0.014445; Mean flowtime = 0.0000

tote = 2826; Entrytime = 195576.9230; Column = 52; Mean throughput = 0.0


tote = 3543; Entrytime = 249814.5972; Column = 32; Mean throughput = 0.014183; Mean flowtime = 0.0000

tote = 3544; Entrytime = 249922.6623; Column = 54; Mean throughput = 0.014180; Mean flowtime = 0.0000

tote = 3545; Entrytime = 250032.4364; Column = 36; Mean throughput = 0.014178; Mean flowtime = 0.0000

tote = 3546; Entrytime = 250068.7008; Column =  1; Mean throughput = 0.014180; Mean flowtime = 0.0000

tote = 3547; Entrytime = 250154.9171; Column = 42; Mean throughput = 0.014179; Mean flowtime = 0.0000

tote = 3548; Entrytime = 250165.4285; Column = 14; Mean throughput = 0.014183; Mean flowtime = 0.0000

tote = 3549; Entrytime = 250278.2364; Column = 31; Mean throughput = 0.014180; Mean flowtime = 0.0000

tote = 3550; Entrytime = 250299.8658; Column =  2; Mean throughput = 0.014183; Mean flowtime = 0.0000

tote = 3551; Entrytime = 250320.7983; Column = 30; Mean throughput = 0.014186; Mean flowtime = 0.0000

tote = 3552; Entrytime = 250452.5619; Column = 12; Mean throughput = 0.0


tote = 4208; Entrytime = 292140.8638; Column = 33; Mean throughput = 0.014404; Mean flowtime = 0.0000

tote = 4209; Entrytime = 292157.8407; Column = 32; Mean throughput = 0.014407; Mean flowtime = 0.0000

tote = 4210; Entrytime = 292264.5331; Column =  2; Mean throughput = 0.014405; Mean flowtime = 0.0000

tote = 4211; Entrytime = 292267.3625; Column = 28; Mean throughput = 0.014408; Mean flowtime = 0.0000

tote = 4212; Entrytime = 292390.4646; Column = 16; Mean throughput = 0.014405; Mean flowtime = 0.0000

tote = 4213; Entrytime = 292493.4665; Column = 44; Mean throughput = 0.014404; Mean flowtime = 0.0000

tote = 4214; Entrytime = 292522.7128; Column = 30; Mean throughput = 0.014406; Mean flowtime = 0.0000

tote = 4215; Entrytime = 292636.9939; Column =  3; Mean throughput = 0.014404; Mean flowtime = 0.0000

tote = 4216; Entrytime = 292712.1407; Column = 50; Mean throughput = 0.014403; Mean flowtime = 0.0000

tote = 4217; Entrytime = 292781.6465; Column = 19; Mean throughput = 0.0


tote = 4816; Entrytime = 335324.9382; Column = 37; Mean throughput = 0.014362; Mean flowtime = 0.0000

tote = 4817; Entrytime = 335353.7076; Column = 17; Mean throughput = 0.014364; Mean flowtime = 0.0000

tote = 4818; Entrytime = 335414.3783; Column = 32; Mean throughput = 0.014364; Mean flowtime = 0.0000

tote = 4819; Entrytime = 335430.6507; Column = 10; Mean throughput = 0.014367; Mean flowtime = 0.0000

tote = 4820; Entrytime = 335499.9988; Column =  4; Mean throughput = 0.014367; Mean flowtime = 0.0000

tote = 4821; Entrytime = 335526.3681; Column = 34; Mean throughput = 0.014368; Mean flowtime = 0.0000

tote = 4822; Entrytime = 335747.2785; Column = 54; Mean throughput = 0.014362; Mean flowtime = 0.0000

tote = 4823; Entrytime = 335790.5179; Column = 29; Mean throughput = 0.014363; Mean flowtime = 0.0000

tote = 4824; Entrytime = 335832.7228; Column = 41; Mean throughput = 0.014364; Mean flowtime = 0.0000

tote = 4825; Entrytime = 335869.5767; Column = 44; Mean throughput = 0.0

tote = 5354; Entrytime = 373938.2748; Column = 43; Mean throughput = 0.014318; Mean flowtime = 0.0000

tote = 5355; Entrytime = 374006.2131; Column = 13; Mean throughput = 0.014318; Mean flowtime = 0.0000

tote = 5356; Entrytime = 374026.4413; Column = 51; Mean throughput = 0.014320; Mean flowtime = 0.0000

tote = 5357; Entrytime = 374045.9278; Column = 30; Mean throughput = 0.014322; Mean flowtime = 0.0000

tote = 5358; Entrytime = 374071.6876; Column = 29; Mean throughput = 0.014323; Mean flowtime = 0.0000

tote = 5359; Entrytime = 374102.4552; Column = 51; Mean throughput = 0.014325; Mean flowtime = 0.0000

tote = 5360; Entrytime = 374164.6247; Column = 25; Mean throughput = 0.014325; Mean flowtime = 0.0000

tote = 5361; Entrytime = 374328.8783; Column =  1; Mean throughput = 0.014322; Mean flowtime = 0.0000

tote = 5362; Entrytime = 374474.6510; Column = 46; Mean throughput = 0.014319; Mean flowtime = 0.0000

tote = 5363; Entrytime = 374521.3125; Column = 46; Mean throughput = 0.01


tote = 5979; Entrytime = 418434.7389; Column = 29; Mean throughput = 0.014289; Mean flowtime = 0.0000

tote = 5980; Entrytime = 418517.5086; Column = 37; Mean throughput = 0.014289; Mean flowtime = 0.0000

tote = 5981; Entrytime = 418518.5953; Column = 45; Mean throughput = 0.014291; Mean flowtime = 0.0000

tote = 5982; Entrytime = 418541.9602; Column = 46; Mean throughput = 0.014292; Mean flowtime = 0.0000

tote = 5983; Entrytime = 418547.2892; Column = 12; Mean throughput = 0.014295; Mean flowtime = 0.0000

tote = 5984; Entrytime = 418547.7701; Column = 30; Mean throughput = 0.014297; Mean flowtime = 0.0000

tote = 5985; Entrytime = 418660.4924; Column = 42; Mean throughput = 0.014296; Mean flowtime = 0.0000

tote = 5986; Entrytime = 418740.4768; Column = 34; Mean throughput = 0.014295; Mean flowtime = 0.0000

tote = 5987; Entrytime = 418768.3460; Column = 17; Mean throughput = 0.014297; Mean flowtime = 0.0000

tote = 5988; Entrytime = 418784.6023; Column = 43; Mean throughput = 0.0


tote = 6653; Entrytime = 466462.7330; Column = 49; Mean throughput = 0.014263; Mean flowtime = 0.0000

tote = 6654; Entrytime = 466500.2885; Column = 50; Mean throughput = 0.014264; Mean flowtime = 0.0000

tote = 6655; Entrytime = 466512.9782; Column = 44; Mean throughput = 0.014265; Mean flowtime = 0.0000

tote = 6656; Entrytime = 466713.9135; Column = 19; Mean throughput = 0.014261; Mean flowtime = 0.0000

tote = 6657; Entrytime = 466746.6770; Column = 48; Mean throughput = 0.014263; Mean flowtime = 0.0000

tote = 6658; Entrytime = 466875.4501; Column =  7; Mean throughput = 0.014261; Mean flowtime = 0.0000

tote = 6659; Entrytime = 466880.1011; Column = 45; Mean throughput = 0.014263; Mean flowtime = 0.0000

tote = 6660; Entrytime = 466884.6512; Column = 25; Mean throughput = 0.014265; Mean flowtime = 0.0000

tote = 6661; Entrytime = 466924.6895; Column =  2; Mean throughput = 0.014266; Mean flowtime = 0.0000

tote = 6662; Entrytime = 467019.8655; Column = 13; Mean throughput = 0.0


tote = 7133; Entrytime = 502016.2571; Column = 48; Mean throughput = 0.014209; Mean flowtime = 0.0000

tote = 7134; Entrytime = 502026.3887; Column = 37; Mean throughput = 0.014210; Mean flowtime = 0.0000

tote = 7135; Entrytime = 502134.0704; Column =  7; Mean throughput = 0.014209; Mean flowtime = 0.0000

tote = 7136; Entrytime = 502213.7538; Column = 54; Mean throughput = 0.014209; Mean flowtime = 0.0000

tote = 7137; Entrytime = 502278.6131; Column = 37; Mean throughput = 0.014209; Mean flowtime = 0.0000

tote = 7138; Entrytime = 502311.9008; Column = 37; Mean throughput = 0.014210; Mean flowtime = 0.0000

tote = 7139; Entrytime = 502356.4400; Column =  4; Mean throughput = 0.014211; Mean flowtime = 0.0000

tote = 7140; Entrytime = 502386.8461; Column =  9; Mean throughput = 0.014212; Mean flowtime = 0.0000

tote = 7141; Entrytime = 502477.9340; Column = 52; Mean throughput = 0.014212; Mean flowtime = 0.0000

tote = 7142; Entrytime = 502555.3738; Column = 15; Mean throughput = 0.0


tote = 7866; Entrytime = 554775.5726; Column = 26; Mean throughput = 0.014179; Mean flowtime = 0.0000

tote = 7867; Entrytime = 554790.4258; Column =  4; Mean throughput = 0.014180; Mean flowtime = 0.0000

tote = 7868; Entrytime = 554845.6539; Column = 46; Mean throughput = 0.014181; Mean flowtime = 0.0000

tote = 7869; Entrytime = 554896.0976; Column = 18; Mean throughput = 0.014181; Mean flowtime = 0.0000

tote = 7870; Entrytime = 554922.7160; Column = 47; Mean throughput = 0.014182; Mean flowtime = 0.0000

tote = 7871; Entrytime = 554947.9667; Column = 11; Mean throughput = 0.014183; Mean flowtime = 0.0000

tote = 7872; Entrytime = 554985.2943; Column =  2; Mean throughput = 0.014184; Mean flowtime = 0.0000

tote = 7873; Entrytime = 555089.5723; Column = 13; Mean throughput = 0.014183; Mean flowtime = 0.0000

tote = 7874; Entrytime = 555093.8508; Column = 13; Mean throughput = 0.014185; Mean flowtime = 0.0000

tote = 7875; Entrytime = 555165.7065; Column = 20; Mean throughput = 0.0


tote = 8348; Entrytime = 591963.8376; Column = 48; Mean throughput = 0.014102; Mean flowtime = 0.0000

tote = 8349; Entrytime = 592013.7164; Column = 53; Mean throughput = 0.014103; Mean flowtime = 0.0000

tote = 8350; Entrytime = 592036.6334; Column = 11; Mean throughput = 0.014104; Mean flowtime = 0.0000

tote = 8351; Entrytime = 592184.6443; Column = 48; Mean throughput = 0.014102; Mean flowtime = 0.0000

tote = 8352; Entrytime = 592244.3646; Column = 15; Mean throughput = 0.014102; Mean flowtime = 0.0000

tote = 8353; Entrytime = 592264.5680; Column = 22; Mean throughput = 0.014103; Mean flowtime = 0.0000

tote = 8354; Entrytime = 592279.5805; Column = 16; Mean throughput = 0.014105; Mean flowtime = 0.0000

tote = 8355; Entrytime = 592280.6927; Column = 23; Mean throughput = 0.014106; Mean flowtime = 0.0000

tote = 8356; Entrytime = 592293.3895; Column =  3; Mean throughput = 0.014108; Mean flowtime = 0.0000

tote = 8357; Entrytime = 592447.7762; Column = 42; Mean throughput = 0.0


tote = 9071; Entrytime = 643002.9607; Column = 18; Mean throughput = 0.014107; Mean flowtime = 0.0000

tote = 9072; Entrytime = 643101.7957; Column = 36; Mean throughput = 0.014107; Mean flowtime = 0.0000

tote = 9073; Entrytime = 643244.8757; Column = 43; Mean throughput = 0.014105; Mean flowtime = 0.0000

tote = 9074; Entrytime = 643356.7929; Column = 13; Mean throughput = 0.014104; Mean flowtime = 0.0000

tote = 9075; Entrytime = 643461.0356; Column = 44; Mean throughput = 0.014103; Mean flowtime = 0.0000

tote = 9076; Entrytime = 643461.2919; Column = 12; Mean throughput = 0.014105; Mean flowtime = 0.0000

tote = 9077; Entrytime = 643563.6699; Column = 16; Mean throughput = 0.014104; Mean flowtime = 0.0000

tote = 9078; Entrytime = 643649.1607; Column = 51; Mean throughput = 0.014104; Mean flowtime = 0.0000

tote = 9079; Entrytime = 643673.0506; Column = 45; Mean throughput = 0.014105; Mean flowtime = 0.0000

tote = 9080; Entrytime = 643694.0384; Column =  8; Mean throughput = 0.0


tote = 9769; Entrytime = 694608.0758; Column = 41; Mean throughput = 0.014064; Mean flowtime = 0.0000

tote = 9770; Entrytime = 694640.8139; Column =  4; Mean throughput = 0.014065; Mean flowtime = 0.0000

tote = 9771; Entrytime = 694699.2398; Column = 44; Mean throughput = 0.014065; Mean flowtime = 0.0000

tote = 9772; Entrytime = 694762.1105; Column =  1; Mean throughput = 0.014065; Mean flowtime = 0.0000

tote = 9773; Entrytime = 694833.1266; Column = 32; Mean throughput = 0.014065; Mean flowtime = 0.0000

tote = 9774; Entrytime = 694836.3729; Column =  4; Mean throughput = 0.014067; Mean flowtime = 0.0000

tote = 9775; Entrytime = 694836.6042; Column = 30; Mean throughput = 0.014068; Mean flowtime = 0.0000

tote = 9776; Entrytime = 694935.8154; Column = 33; Mean throughput = 0.014067; Mean flowtime = 0.0000

tote = 9777; Entrytime = 694953.2362; Column =  5; Mean throughput = 0.014069; Mean flowtime = 0.0000

tote = 9778; Entrytime = 695012.9109; Column = 48; Mean throughput = 0.0

c. Confirm that your simulations produce the correct throughput and flow time.

## 6.3 Exercise 3: Vehicle as machine (5 points)
Next, we include the Vehicle which as a first attempt is modelled as a standard machine with a deterministic processing time of 40.0 seconds

a. Determine (analytically) the throughput and flow time that should result from your model with only the generator process, demand buffer, vehicle and the exit process.

b. Complete the simulation code in the template below.

Note: The previously defined variables and functions remain unchanged, only the Vehicle is added, and the main code is changed.

In [ ]:
# =================================
# # Vehicle definition
# =================================
def Vehicle(env, receiving_channel, sending_channel):
    ### BEGIN SOLUTION
    tp = 40.0  # processing time
    while True:
        receive = receiving_channel.receive()
        x = yield env.select(receive)
        yield env.timeout(tp) 
        send = sending_channel.send(x)
        yield env.select(send)
    ### END SOLUTION
        
# =================================
# # Main
# =================================
env = Environment()
a = Channel(env)
b = Channel(env)
c = Channel(env)
G = env.process(Generator(env, a))
D = env.process(Demand_Buffer(env, a,b))
V = env.process(Vehicle(env, b,c))
E = env.process(Exit(env, c))
env.run(until=E)
print ("simulation has ended")

c. Confirm that your simulations produce the correct throughput and flow time.

d. Run your simulation at least 10 times. What is the range of outcomes for the mean throughput? Determine (roughly/ by trial and error) the number of orders required to get good estimates for the flow time (first three non-zero digits are correct).

## 6.4 Exercise 4: Vehicle accurately modelled (15 points)
Assume that a vehicle starts from the buffer. If it has to go to column $i \in \{0, depth - 1\}$, it needs to drive a distance $dv \cdot (i + 1)$, pick up the tote, drive back, and deliver the tote to the buffer. For driving the required distance, the vehicle first goes with a constant acceleration `av` *m/s$^2$* until it reaches its maximum velocity `vmaxv`, then it moves with this velocity and decelerates with a constant deceleration `av` to approach its destination point. If the tote is located too close to the end of the aisle, it is possible that the vehicle will not reach its maximum speed. 

a. From this information, (analytically) determine the time (in seconds) it takes a vehicle to drive a given distance of x meters. To that end, first determine from the acceleration profile the travelled distance as function of time. From that you can determine the time required to travel a given distance.

b. Next, if loading and unloading a tote both take `lv` seconds, determine (analytically) the time it takes a vehicle, starting from the buffer, to pick up a tote from column $i \in \{0, depth\}$ and deliver it to the buffer.

c. Determine (analytically) the mean and variance of this total travel time for a random (uniformly distributed) column.

d. Determine (analytically) the throughput and flow time that should result from your model with only the generator process, demand buffer, vehicle and the exit process.

e. Complete the simulation code in the template below, incorporating the actual time a vehicle requires for picking up a tote and delivering it to the buffer.

Note: the previously defined variables and functions remain unchanged, only the Vehicle needs to be redefined.

<span style="color:red">(Ik ben het niet helemaal eens met hoe de vehicle gemodelleerd wordt.)</span>

In [ ]:
# =================================
# # Global vehicle variables
# =================================
lv = 3.0                  # time to load/unload the vehicle
dv = 0.5                  # unit width clearance
vmaxv = 1.5              # maximum velocity of the vehicle
av = 1.0                  # acceleration/deceleration of the vehicle

# =================================
# # Vehicle definition
# =================================
def Vehicle(env, receiving_channel, sending_channel):
    ### BEGIN SOLUTION
    while True: 
        receive = receiving_channel.receive()
        x = yield env.select(receive)
        
        distance = dv*x.column
        if distance <= vmaxv**2/av:
            t_travel = math.sqrt(4*distance/av)
        else:
            t_travel = 2*vmaxv/av  +  (distance-vmaxv**2/av)/vmaxv
        
        # TODO: eigenlijk zou je hier pas het item kunnen ophalen, maar zo wordt het niet gemodelleerd.
        # De reden is natuurlijk dat je in dat geval geen idee hebt waar je naartoe moet rijden. Dit zou eigenlijk eerst verstuurd moeten worden.
        #  x = yield receiving_channel.receive()
        # TODO: Daarnaast, zou je pas kunnen unloaden als er een plekje vrij is in de buffer.
        # Dus pas als de buffer ready is om te ontvangen, heb je nog lv seconden nodig om het ook echt in de buffer te plaatsen.
        #  yield env.timeout(t_travel)
        #  yield sending_channel.readyto_send()
        #  yield env.timeout(lv)
        #  yield sending_channel.send(x)
        te_processing = 2*t_travel+2*lv
        yield env.timeout(te_processing)
        
        send = sending_channel.send(x)  
        yield env.select(send)    
    ### END SOLUTION

# =================================
# # Main
# =================================
env = Environment()
a = Channel(env)
b = Channel(env)
c = Channel(env)
G = env.process(Generator(env, a))
D = env.process(Demand_Buffer(env, a,b))
V = env.process(Vehicle(env, b,c))
E = env.process(Exit(env, c))
env.run(until=E)
print ("simulation has ended")


f. Confirm that your simulations produce the correct throughput and flow time.

g.  Run your simulation at least 10 times. What is the range of outcomes for the mean throughput? Determine (roughly, by trial and error) the number of orders required to get good estimates for the flow time (first three non-zero digits are correct).

## 6.5 Exercise 5: Lift as two stage machine (5 points)
In the previous exercises we focussed on modeling a tier. In exercises 5-8 we focus on modeling the Buffer and Lift, starting with the Lift. We model the Lift as a machine, which repeatedly does the following: request a destination tier, move to the desired tier, receive a tote (first delay for `ll`, then communicate), move to the ground floor, deliver the tote (first delay for `ll`, then communicate). In this exercise we assume that moving to the desired tier takes 23.0 seconds, receiving the tote takes 2.0 seconds, moving to the ground floor again takes 23.0 seconds, and delivering the tote also takes 2.0 seconds. Note that, in this case, this implies that the tote should leave the Generator and enter the Lift 25.0 seconds after the Lift has received the destination tier. All of these times are assumed to be deterministic.

a. Assume that after 1.0 second the first tote arrives. Subsequently a tote arrives respectively 10.0, 100.0, and 100.0 seconds after the completion of sending its predecessor to the buffer. Make a lot-time-diagram indicating how long a newly generated job spends in the Generator, and spends in the Lift.

b. Complete the simulation code in the template below, incorporating the actual time a vehicle requires for picking up a tote and delivering it to the buffer.

Note: we first reinitalize the entire simulation, as changes are made in the entire simulation model (generator, lift, exit, etc.)

In [1]:
%reset -f
from numpy import random
import simpy
import math
import numpy
numpy.seterr(divide='ignore')
from Course_4DC10.simChi.core.channel_select import Channel, process, selected, Environment
from dataclasses import dataclass



SimChi version 20210730 imported succesfully.
 


In [18]:
# =================================
# # Tote definition
# =================================
@dataclass
class Tote:
    entrytime: float = 0
    tier: int = 0

# =================================
# # Global Lift variables
# =================================
ll = 2.0  # time to load/unload the lift


# =================================
# # Generator definition
# =================================
def Generator(env, sending_channel, call_lift_channel):
    delays=[1.0, 10.0, 100.0, 100.0]
    while len(delays)>0:
        delay = env.timeout(delays[0])
        yield delay
        delays = delays[1:]
        
        send = call_lift_channel.send(1)
        yield env.select(send)
        print("Generator: Informed Lift to go to tier 1 at time %3.1f" % env.now);
        # TODO? De tijd wordt bepaald op het moment dat je gaat wachten
        # Eigenlijk moet de tijd bepaald worden op het moment dat je echt verstuurd!
        # Als je x daarvoor definieert, dan krijg je een andere entry time 
        # dan als je x definieert op het moment van sturen.
        ## Dit geldt ook voor alle andere exercises
        x = Tote()
        x.entrytime = env.now
        x.tier = 1 
        send = sending_channel.send(x)
        yield env.select(send)
        print("Generator: Tote has completely left the Generator and entered the Lift at time %3.1f" % env.now)
    delay = env.timeout(100.0)
    yield delay


# =================================
# # Lift definition
# =================================
def Lift(env, receiving_channel, call_lift_channel, sending_channel):
    ### BEGIN SOLUTION
    while True:
        lift_called = call_lift_channel.receive()
        yield env.select(lift_called)
        
        t_travel = 23.0
        yield env.timeout(t_travel+ll)
        
        receive = receiving_channel.receive()       
        yield env.select(receive)
        x = receive.received_entity
        
        yield env.timeout(t_travel+ll)
        
        send = sending_channel.send(x)
        yield env.select(send)
    ### END SOLUTION

    
# =================================
# # Exit definition
# =================================
def Exit(env, receiving_channel):
    while True:
        receive = receiving_channel.receive()
        x = yield env.select(receive)
        print("Exit:      Tote has completely left the Lift and has been received by the Exit at time %3.1f" % env.now);


# =================================
# # Main
# =================================
env = Environment()
a = Channel(env)  # for sending lots
b = Channel(env)  # lift tier
c = Channel(env)  # for sending lots
G = env.process(Generator(env, a, b))
L = env.process(Lift(env, a, b, c))
E = env.process(Exit(env, c))
env.run()
print ("simulation has ended")



Generator: Informed Lift to go to tier 1 at time 1.0
Generator: Tote has completely left the Generator and entered the Lift at time 26.0
Exit:      Tote has completely left the Lift and has been received by the Exit at time 51.0
Generator: Informed Lift to go to tier 1 at time 51.0
Generator: Tote has completely left the Generator and entered the Lift at time 76.0
Exit:      Tote has completely left the Lift and has been received by the Exit at time 101.0
Generator: Informed Lift to go to tier 1 at time 176.0
Generator: Tote has completely left the Generator and entered the Lift at time 201.0
Exit:      Tote has completely left the Lift and has been received by the Exit at time 226.0
Generator: Informed Lift to go to tier 1 at time 301.0
Generator: Tote has completely left the Generator and entered the Lift at time 326.0
Exit:      Tote has completely left the Lift and has been received by the Exit at time 351.0
simulation has ended


c. Confirm that your simulations produce the correct results, i.e., in correspondence with your lot-time-diagram.

## 6.6 Exercise 6: Buffer with only one tier (5 points)
In case we have only one tier, we could use a standard model for a finite buffer. However, keeping in mind that we need to model in the next exercise Level finite buffers in parallel where totes are served at a first-come-first-serve basis, we model the finite buffer slightly differently. We have a counter `n` which counts the number of jobs in the buffer, and we keep a list `xs` of totes that have arrived to the buffer. As long as the buffer contains less than `bc` totes, the buffer can receive jobs. Furthermore, if the buffer contains totes, it is possible to receive a request from the Lift (asking for which tier to move to, which is always 0 in this case). Additionally, if the buffer contains totes it is possible to send the first tote of the list `xs` to the lift.

a. Assume that after 1.0 second the first tote arrives. Subsequently a tote arrives respectively 10.0, 100.0, and 100.0 seconds after the completion of sending its predecessor to the buffer. Make a lot-time-diagram indicating how long a newly generated job spends in the Generator, in the Buffer, and spends in the Lift.

 b. Complete the simulation code in the template below, only the Generator and Buffer need to be redefined.

In [19]:
# =================================
# # Global Lift variables
# =================================
ll = 2.0  # time to load/unload the lift
bc = 1    # buffer capacity

# =================================
# # Generator definition
# =================================
def Generator(env, sending_channel):
    delays=[1.0, 10.0, 100.0, 100.0]
    while len(delays)>0:
        delay = env.timeout(delays[0])
        yield delay
        delays = delays[1:]
        
        x = Tote()
        x.entrytime = env.now
        x.tier = 0
        send = sending_channel.send(x)
        yield env.select(send)
        print("Generator: sending tote to Buffer at time %3.1f" % env.now)
    delay = env.timeout(100.0)
    yield delay

# =================================
## Buffer definition
# =================================
def Buffer(env, receiving_channel, sending_channel, call_lift_channel):
    xs = [] # list of jobs
    n = 0
    while True:
        receiving = receiving_channel.receive() if n < bc else None
        call_lift = call_lift_channel.send(xs[0].tier) if len(xs) > 0 else None
        sending   = sending_channel.send(xs[0]) if len(xs)>0 else None
        
        yield env.select(receiving, call_lift, sending)

        if selected(receiving):
            x = receiving.received_entity
            xs = xs + [x]
            n = n+1
        
        if selected(call_lift):
            pass
        
        if selected(sending):
            ### BEGIN SOLUTION
            xs = xs[1:]
            n = n - 1
            ### END SOLUTION
            
# =================================
## Main
# =================================
env = Environment()
a = Channel(env)  # for sending lots
b = Channel(env)  # for sending lots
c = Channel(env)  # for calling the lift
d = Channel(env)  # for sending lots

G = env.process(Generator(env, a))
B = env.process(Buffer(env, a, b, c))
L = env.process(Lift(env, b, c, d))
E = env.process(Exit(env, d))
env.run()
print ("simulation has ended")

Generator: sending tote to Buffer at time 1.0
Generator: sending tote to Buffer at time 26.0
Exit:      Tote has completely left the Lift and has been received by the Exit at time 51.0
Exit:      Tote has completely left the Lift and has been received by the Exit at time 101.0
Generator: sending tote to Buffer at time 126.0
Exit:      Tote has completely left the Lift and has been received by the Exit at time 176.0
Generator: sending tote to Buffer at time 226.0
Exit:      Tote has completely left the Lift and has been received by the Exit at time 276.0
simulation has ended


c. Confirm that your simulations produce the correct results, i.e., in correspondence with your lot-time-diagram

## Exercise 7: Buffer with multiple tiers (5 points)
We extend the buffer from the previous exercise to `Level` parallel finite buffers of size `bc` which are modelled as a list of `Level` counters. For this list we use the variable `n`. We keep a common list `xs` of totes that have arrived to all buffers. This list is used to determine which tote should be picked up first by the Lift. For each of the tiers it holds that if the buffer of that tier is not full yet, a tote can be received. Furthermore, if the buffer contains totes, it is possible to receive a request from the Lift asking for the tier to move to. Additionally, if the buffer contains totes it is possible to send the first tote of the list `xs` to the lift.

a. Consider two tiers, and assume that after `Tier` seconds the first tote arrives (for two tiers: after 0.0 for tier 0 and
after 1.0 for tier 1). Subsequently a tote arrives respectively 10.0, 100.0, and 100.0 seconds after the completion
of sending its predecessor to the buffer. Make a lot-time-diagram indicating how long a newly generated job
spends in which Generator, in the Buffer, and spends in the Lift.

b. Complete the simulation code in the template below.

In [20]:
from Course_4DC10.simChi.core.channel_select import Channel, process, selected, Environment
# =================================
# # Global variables
# =================================
ll = 2.0  # time to load/unload the lift
bc = 1    # buffer capacity
Level = 2 # the number of tiers

# =================================
# # Generator definition
# =================================
def Generator(env, sending_channel, level):
    delays=[1.0*level, 10.0, 100.0, 100.0]
    while len(delays)>0:
        delay = env.timeout(delays[0])
        yield delay
        delays = delays[1:]
        
        x = Tote()
        x.entrytime = env.now
        x.tier = level
        send = sending_channel.send(x)
        yield env.select(send)
        print("Generator %d: Sending tote to Buffer completed at time %3.1f" % (level, env.now) )
    delay = env.timeout(100.0)
    yield delay 
    
# =================================
# # Buffer definition
# =================================
def Buffer(env, receiving_channels, sending_channel, call_lift_channel):
    xs = []
    n = [0] * Level
    while True:
        receiving = [receiving_channels[Tier].receive() if n[Tier] < bc else None for Tier in range(Level) ]
        call_lift = call_lift_channel.send(xs[0].tier) if len(xs) > 0 else None
        sending = sending_channel.send(xs[0]) if len(xs) > 0 else None
        communications = receiving + [call_lift] + [sending]
        yield env.select(*communications)
        
        for Tier in range(Level):
            if selected(receiving[Tier]):
                x = receiving[Tier].received_entity
                xs = xs + [x]
                n[Tier] = n[Tier] + 1
        
        if selected(call_lift):
            pass
            
        if selected(sending):
            ### BEGIN SOLUTION
            Tier = xs[0].tier 
            xs = xs[1:]
            n[Tier] = n[Tier] - 1
            ### END SOLUTION
            
# =================================
# # Main
# =================================
env = Environment()
a = [Channel(env) for tier in range(Level)] # a channel for each tier, each sending lots
b = Channel(env)  # for sending lots
c = Channel(env)  # for calling the lift
d = Channel(env)  # for sending lots
Gs = [env.process(Generator(env, a[tier], tier)) for tier in range(Level)] 
B = env.process(Buffer(env, a, b, c))
L = env.process(Lift(env, b, c, d))
E = env.process(Exit(env, d))
env.run(until=1000000)
print ("simulation has ended")

Generator 0: Sending tote to Buffer completed at time 0.0
Generator 1: Sending tote to Buffer completed at time 1.0
Generator 0: Sending tote to Buffer completed at time 25.0
Exit:      Tote has completely left the Lift and has been received by the Exit at time 50.0
Generator 1: Sending tote to Buffer completed at time 75.0
Exit:      Tote has completely left the Lift and has been received by the Exit at time 100.0
Generator 0: Sending tote to Buffer completed at time 125.0
Exit:      Tote has completely left the Lift and has been received by the Exit at time 150.0
Generator 1: Sending tote to Buffer completed at time 175.0
Exit:      Tote has completely left the Lift and has been received by the Exit at time 200.0
Generator 0: Sending tote to Buffer completed at time 225.0
Exit:      Tote has completely left the Lift and has been received by the Exit at time 250.0
Generator 1: Sending tote to Buffer completed at time 275.0
Exit:      Tote has completely left the Lift and has been rece

c. Confirm that your simulations produce the correct results, i.e., in correspondence with your lot-time-diagram

## Exercise 8: Lift accurately modelled (5 points)

Assume that the Lift starts from the ground floor. If it has to go to level $i \in \{0, Level - 1\}$, it needs to travel a
distance $dl \cdot (i + 1)$, pick up the tote, travel back, and deliver the tote at the ground floor. For traveling the required
distance, the vehicle first goes with a constant acceleration `al` *m/s$^2$* until it reaches its maximum velocity `vmaxl`, then
it moves with this velocity and decelerates with a constant deceleration `al` to approach its destination tier. If the tier
is located too close to the ground floor, it is possible that the Lift will not reach its maximum speed.

a. From this information, (analytically) determine the time (in seconds) it takes the Lift to a given tier.

b. Next, if loading and unloading a tote both take `ll` seconds, determine (analytically) the time it takes the Lift to pick up a tote from tier $i \in \{0, depth - 1\}$ and deliver it to the buffer.

c. Consider two tiers, and assume that after Tier seconds the first tote arrives (for two tiers: after 0.0 for tier 0 and after 1.0 for tier 1). Subsequently a tote arrives respectively 10.0, 100.0, and 100.0 seconds after the completion of sending its predecessor to the buffer. Make a lot-time-diagram indicating how long a newly generated job spends in which Generator, in the Buffer, and spends in the Lift.

d. Complete the simulation code in the template below.

In [22]:
# =================================
# # Global variables
# =================================
ll = 2.0        # time to load/unload the lift
bc = 1          # buffer capacity
Level = 2       # the number of tiers
dl = 0.8        # unit height clearance
vmaxl = 5.0     # maximum velocity of lift
al = 7.0        # acceleration/deceleration of lift

# =================================
# # Lift definition
# =================================
def Lift(env, receiving_channel, call_lift_channel, sending_channel):
    while True: 
        ### BEGIN SOLUTION
        lift_called = call_lift_channel.receive()
        yield env.select(lift_called)
        requested_tier = lift_called.received_entity
        
        distance = dl * requested_tier
        if distance <= vmaxl**2/al:
            t_travel = math.sqrt(4*distance/al)
        else:
            t_travel = 2*vmaxl/al  +  (distance-vmaxl**2/al)/vmaxl
        yield env.timeout(t_travel+ll)
        
        receive = receiving_channel.receive()       
        yield env.select(receive)
        x = receive.received_entity
        
        yield env.timeout(t_travel+ll)
        
        send = sending_channel.send(x)
        yield env.select(send)
        ### END SOLUTION
        
        
# =================================
# # Main
# =================================
env = Environment()
a = [Channel(env) for Tier in range(Level)] # a channel for each tier, each sending totes
b = Channel(env)  # for sending totes
c = Channel(env)  # for calling the lift
d = Channel(env)  # for sending totes

Gs = [env.process(Generator(env, a[Tier], Tier)) for Tier in range(Level)] 
B = env.process(Buffer(env, a, b, c))
L = env.process(Lift(env, b, c, d))
E = env.process(Exit(env, d))
env.run()
print ("simulation has ended")

Generator 0: Sending tote to Buffer completed at time 0.0
Generator 1: Sending tote to Buffer completed at time 1.0
Exit:      Tote has completely left the Lift and has been received by the Exit at time 4.0
Exit:      Tote has completely left the Lift and has been received by the Exit at time 9.4
Generator 0: Sending tote to Buffer completed at time 10.0
Generator 1: Sending tote to Buffer completed at time 11.0
Exit:      Tote has completely left the Lift and has been received by the Exit at time 14.0
Exit:      Tote has completely left the Lift and has been received by the Exit at time 19.4
Generator 0: Sending tote to Buffer completed at time 110.0
Generator 1: Sending tote to Buffer completed at time 111.0
Exit:      Tote has completely left the Lift and has been received by the Exit at time 114.0
Exit:      Tote has completely left the Lift and has been received by the Exit at time 119.4
Generator 0: Sending tote to Buffer completed at time 210.0
Generator 1: Sending tote to Buffe

e. Confirm that your simulations produce the correct results, i.e., in correspondence with your lot-time-diagram

## 6.9 Exercise 9: Entire system (10 points)
In the previous exercises we developed and tested each process individually. Now it is time to combine all processes into one model.

a. Determine the (total) arrival rate of orders for a system with `Level` tiers, where for each tier orders arrive with a mean inter arrival time of `arrive`.

b. Complete the simulation code in the template below by substituting the code for the processes as you derived in Exercises 1, 2, 4, 7, and 8. Note that the type tote has been extended in comparison with Exercise 1, so one line of code extra is required in the Generator process to include the tier.

In [14]:
import time
start_time = time.time()

# =================================
# # Tote definition
# =================================
@dataclass
class Tote:
    entrytime: float = 0.0
    column: int = 0
    tier: int  = 0
        
# =================================
## Global variables
# =================================
lv = 3.0                   # time to load/unload the vehicle
dv = 0.5                   # unit width clearance
vmaxv = 1.5                # maximum velocity of the vehicle
av = 1.0                   # acceleration/deceleration of the vehicle
ll = 2.0                   # time to load/unload the lift
dl = 0.8                   # unit height clearance
vmaxl = 5.0                # maximum velocity of lift
al = 7.0                   # acceleration/deceleration of lift
bc = 1                     # buffer capacity
arrive = 70.0              # inter arrival time of requests
Level = 9                  # the number of tiers
depth  = 55                # the number of columns
number_of_orders = 10000  # the number of orders to process

# =================================
# # Generator definition
# =================================
def Generator(env, sending_channel, Tier):
    ### BEGIN SOLUTION
    ta_dist = lambda: random.exponential(arrive)
    column_dist = lambda:  random.randint(1, depth)
    ### END SOLUTION
    while True:
        x = Tote()
        x.entrytime = env.now
        # x.column = ..... 
        # x.tier = ..... 
        ### BEGIN SOLUTION
        x.column = column_dist()
        x.tier = Tier
        ### END SOLUTION
        send = sending_channel.send(x)
        yield env.select(send)
        # delay = ..... 
        ### BEGIN SOLUTION
        delay = env.timeout(ta_dist())
        ### END SOLUTION
        yield delay
        
        
# =================================
# # Demand Buffer definition
# =================================
def Demand_Buffer(env, receiving_channel, sending_channel):
    xs = [] # list of totes
    while True:
        ### BEGIN SOLUTION
        receiving = receiving_channel.receive() 
        sending = sending_channel.send(xs[0]) if len(xs)>0 else None
        yield env.select(sending,receiving)
        if selected(sending):
            xs = xs[1:]
        if selected(receiving):
            x = receiving.received_entity
            xs = xs + [x]
        ### END SOLUTION
            
            
# =================================
# # Vehicle definition
# =================================
def Vehicle(env, receiving_channel, sending_channel):
    ### BEGIN SOLUTION
    while True: 
        receive = receiving_channel.receive()
        x = yield env.select(receive)
        
        distance = dv*x.column
        if distance <= vmaxv**2/av:
            t_travel = math.sqrt(4*distance/av)
        else:
            t_travel = 2*vmaxv/av  +  (distance-vmaxv**2/av)/vmaxv
        
        # TODO: eigenlijk zou je hier pas het item kunnen ophalen, maar zo wordt het niet gemodelleerd.
        # De reden is natuurlijk dat je in dat geval geen idee hebt waar je naartoe moet rijden. Dit zou eigenlijk eerst verstuurd moeten worden.
        #  x = yield receiving_channel.receive()
        # TODO: Daarnaast, zou je pas kunnen unloaden als er een plekje vrij is in de buffer.
        # Dus pas als de buffer ready is om te ontvangen, heb je nog lv seconden nodig om het ook echt in de buffer te plaatsen.
        #  yield env.timeout(t_travel)
        #  yield sending_channel.readyto_send()
        #  yield env.timeout(lv)
        #  yield sending_channel.send(x)
        te_processing = 2*t_travel+2*lv
        yield env.timeout(te_processing)
        
        send = sending_channel.send(x)  
        yield env.select(send)    
    ### END SOLUTION
    
# =================================
# # Buffer definition
# =================================
def Buffer(env, receiving_channels, sending_channel, call_lift_channel):
    xs = []
    n = [0] * Level
    while True:
        receiving = [receiving_channels[Tier].receive() if n[Tier] < bc else None for Tier in range(Level) ]
        call_lift = call_lift_channel.send(xs[0].tier) if len(xs) > 0 else None
        sending = sending_channel.send(xs[0]) if len(xs) > 0 else None
        communications = receiving + [call_lift] + [sending]
        yield env.select(*communications)
        
        for Tier in range(Level):
            if selected(receiving[Tier]):
                x = receiving[Tier].received_entity
                xs = xs + [x]
                n[Tier] = n[Tier] + 1
        
        if selected(call_lift):
            pass
            
        if selected(sending):
            ### BEGIN SOLUTION
            Tier = xs[0].tier 
            xs = xs[1:]
            n[Tier] = n[Tier] - 1
            ### END SOLUTION
            
# =================================
# # Lift definition
# =================================
def Lift(env, receiving_channel, call_lift_channel, sending_channel):
    while True: 
        ### BEGIN SOLUTION
        lift_called = call_lift_channel.receive()
        yield env.select(lift_called)
        requested_tier = lift_called.received_entity
        
        distance = dl * requested_tier
        if distance <= vmaxl**2/al:
            t_travel = math.sqrt(4*distance/al)
        else:
            t_travel = 2*vmaxl/al  +  (distance-vmaxl**2/al)/vmaxl
        yield env.timeout(t_travel+ll)
        
        receive = receiving_channel.receive()       
        yield env.select(receive)
        x = receive.received_entity
        
        yield env.timeout(t_travel+ll)
        
        send = sending_channel.send(x)
        yield env.select(send)
        ### END SOLUTION
        
# =================================
# # Exit definition
# =================================
def Exit(env, receiving_channel):
    mphi = 0.0
    timer = time.time()
    for i in range(1, number_of_orders + 1):
        receive = receiving_channel.receive()
        yield env.select(receive)
        x = receive.received_entity
    
        mphi = (i - 1) / i * mphi + (env.now - x.entrytime) / i
#         print("tote = %d; Entrytime = %10.4f; Column = %2d; Mean throughput = %8.6f; Mean flowtime = %6.4f\n" %
#              (i , x.entrytime, x.column, i / env.now, mphi ) )

        # Calculate speed of simulation
        if i % 1000 == 0:
            print("--- Totes %s - %s took %s seconds ---" % (i-1000,i, time.time() - timer))
            timer = time.time()
            
        
# =================================
## GDV definition
# =================================
def GDV(env, sending_channel, Tier):
    a = Channel(env)  # for sending totes
    b = Channel(env)  # for sending totes
    G = env.process(Generator(env, a, Tier))
    D = env.process(Demand_Buffer(env, a,b))
    V = env.process(Vehicle(env, b, sending_channel))
        

# =================================
# # Main
# =================================
env = Environment()
c = [Channel(env) for Tier in range(Level)] # a channel for each tier, each sending totes
d = Channel(env)  # for sending totes
e = Channel(env)  # for calling the lift
f = Channel(env)  # for sending totes
GDVs = [ GDV(env, c[Tier], Tier)  for Tier in range(Level) ] 
B = env.process(Buffer(env, c, d, e))
L = env.process(Lift(env, d, e, f))
E = env.process(Exit(env, f))
env.run(until=E)
print ("simulation has ended")

print("--- %s seconds ---" % (time.time() - start_time))

--- Totes 0 - 1000 took 3.140777111053467 seconds ---
--- Totes 1000 - 2000 took 6.032609224319458 seconds ---
--- Totes 2000 - 3000 took 10.749584674835205 seconds ---
--- Totes 3000 - 4000 took 17.1526837348938 seconds ---
--- Totes 4000 - 5000 took 18.595486879348755 seconds ---
--- Totes 5000 - 6000 took 20.930529832839966 seconds ---
--- Totes 6000 - 7000 took 23.45594835281372 seconds ---
--- Totes 7000 - 8000 took 31.634292125701904 seconds ---
--- Totes 8000 - 9000 took 33.76323699951172 seconds ---
--- Totes 9000 - 10000 took 37.076666831970215 seconds ---
simulation has ended
--- 202.53975176811218 seconds ---


c. Verify if the throughput in your simulations matches with the start rate you determined. Verify that totes reach the exit from each tier. Also verify that totes are generated from each column.

d. Run your simulation 30 times, and record for each run the resulting average throughput and average flow time.
Determine the mean and the standard deviation of both the average throughput and the average flow time for these 30 simulations. Finally, determine a 95% confidence interval for both the average throughput and the average flow time.

## 6.10 Exercise 10: Effect of buffer capacity (10 points)
Via computer simulation analyze the impact of the buffer capacity on the throughput and flow time (choose values of `bc` as 1,2,3,200). For each setting, run your simulations 30 times and determine both mean and standard deviation of the resulting average throughput and average flow time for these 30 simulations, as well as 95% confidence intervals.

## 6.11 Exercise 11: Looking for the best layout (30 points)
Via computer simulation consider 6 different scenarios (see the table below) that correspond to different layouts of the
storage facility with a total capacity of 500 totes with different number of tiers and columns.

|  |  |  |  |  | |  |
| - | - | - | - | - | - | - |
| Level |  2 | 5 | 10 | 20 | 25 | 50 |
| Depth | 250 | 100 | 50 | 25 | 20 | 10 |

In order to obtain an answer to the questions below: for each scenario, run your simulations 30 times and determine both mean and standard deviation of the resulting average throughput and average flow time for these 30 simulations, as well as 95% confidence intervals.

a. For each scenario, determine by means of simulation the maximal throughput.

b. Assume a required throughput of 6 totes per minute. Compare the resulting flow time for each feasible scenario.

c. Based on your outcomes, select a layout that seems the best for you. If necessary, introduce other performance indicators and modify the code accordingly. The report on this step should contain an argument to select the best layout.

## References
[1] G. Marchet, M. Melanci, S. Perotti and E. Tappia, "Analytical model to estimate performance of autonomous vehicle storage and retrieval systems for product totes", International Journal of Production Research, vol. 50(24), 7134-7148, 2012.